### ISSUE TITLE
Installation fails with Python 3.7

### ISSUE THREAD

**Issue by wwaites** [bitextor/bitextor#90]
Capturing here the report sent in email to the paracrawl list.

To try to duplicate this as well as familiarise myself with the code and pipeline,
I'm just installed bitextor on my GPU workstation. It is running Ubuntu Linux LTS
18.04. I generally avoid using the system python, instead using miniconda to get a
self-contained environment. The preamble to the process is:

    conda create -n paracrawl python=3
    conda activate paracrawl

Ultimately, it would be ideal if we could just do "conda install bitextor".

This results in an environment with Python 3.7.3.

Beginning the installation,

    pip install -r requirements.txt

First problem is it is unhappy with the nailed tensorflow version:

    ERROR: Could not find a version that satisfies the requirement tensorflow==1.12.1 (from -r requirements.txt (line 12)) (from versions: 1.13.0rc1, 1.13.0rc2, 1.13.1,
2.0.0a0)

I move forward by removing the version constraint from requirements.txt and doing

    conda install tensorflow-gpu

This is nice because it installs all of the external dependencies, CUDA libraries
and whatnot included. I hope that there is not an incompatibility between tensorflow
1.12.1 and 1.13.1 that will cause a problem.

Re-running the `pip install` proceeds until the next problem, which is with sklearn.
It seems to have a similar problem as with datrie. Removing the version constraint
allows this to continue and I end up with 0.20.0 instead of 0.19.1.

As Amir mentioned, datrie also fails, but

    conda install datrie
gets a working version.

The next problem is with the requests library:

    ERROR: requests 2.21.0 has requirement urllib3<1.25,>=1.21.1, but you'll have urllib3 1.25.2 which is incompatible.

This one is solved by adding the constraint to requrements.txt:

    urllib3<1.25

The remaining problems, as found using `pip check`, relate to tensorflow versions:

    tensorflow 1.13.1 has requirement keras-applications>=1.0.6, but you have keras-applications 1.0.2.
    tensorflow 1.13.1 has requirement keras-preprocessing>=1.0.5, but you have keras-preprocessing 1.0.1.

This is a little annoying, but is solved by uninstalling keras (2.2.0) and re-installing
it just using `pip install` which gets (2.2.4).

The specific versions of everything that has been installed by this message are in
the attached text file (output of `pip freeze`) now that everything *thinks* that
it has compatible dependencies.

Continuing,

    pip install -r bicleaner/requirements.txt https://github.com/vitaka/kenlm/archive/master.zip

This needs to remove the version constraints on scikit-learn, numpy
and scipy, for similar reasons to the previous message.

The rest of the installation (autogen.sh && make) seems to work fine.

**Comment by wwaites** [comment]
```
absl-py==0.7.1
alcazar==0.4.0
appdirs==1.4.3
astor==0.7.1
atomicwrites==1.3.0
attrs==19.1.0
beautifulsoup4==4.7.1
boilerpipe3==1.1
cchardet==2.1.4
Cerberus==1.2
certifi==2019.3.9
charade==1.0.3
chardet==3.0.4
ConfigArgParse==0.14.0
cssselect==1.0.3
cycler==0.10.0
datrie==0.7.1
docopt==0.6.2
docutils==0.14
ftfy==5.5.0
func-timeout==4.3.0
gast==0.2.2
gitdb2==2.0.5
GitPython==2.1.11
grpcio==1.16.1
h5py==2.8.0
html2text==2018.1.9
html5lib==1.0.1
idna==2.8
iso-639==0.4.5
jmespath==0.9.4
joblib==0.13.2
JPype1-py3==0.5.5.2
jsonschema==3.0.1
kenlm==0.0.0
Keras==2.2.4
Keras-Applications==1.0.7
Keras-Preprocessing==1.0.9
kiwisolver==1.1.0
langid==1.1.6
lxml==4.2.5
Markdown==3.1
matplotlib==2.2.2
mkl-fft==1.0.12
mkl-random==1.0.2
mock==2.0.0
more-itertools==7.0.0
mosestokenizer==1.0.0
nltk==3.3
numpy==1.16.3
openfile==0.0.7
pbr==5.1.3
pluggy==0.11.0
protobuf==3.7.1
py==1.8.0
pycld2==0.31
pyparsing==2.4.0
pyrsistent==0.15.1
pytest==3.8.1
python-dateutil==2.7.5
python-Levenshtein==0.12.0
python-magic==0.4.15
pytz==2019.1
PyYAML==4.2b4
ratelimiter==1.2.0.post0
regex==2018.1.10
requests==2.21.0
requests-file==1.4.3
sacrebleu==1.2.12
scikit-learn==0.21.0
scipy==1.2.1
six==1.12.0
sklearn==0.0
smmap2==2.0.5
snakemake==5.4.2
soupsieve==1.9.1
tensorboard==1.13.1
tensorflow==1.13.1
tensorflow-estimator==1.13.0
termcolor==1.1.0
tld==0.9.1
tldextract==2.2.0
toolwrapper==0.3.0
tqdm==4.28.1
typing==3.6.6
urllib3==1.24.3
warc3-wet==0.2.3
wcwidth==0.1.7
webencodings==0.5.1
Werkzeug==0.15.2
wrapt==1.11.1
```

**Comment by lpla** [comment]
Hello William.

It is great to hear that you are testing bitextor in another environment like conda. I would also love to have a single command way to install bitextor too, with docker, conda, pip or apt. But sadly it is not a priority now given the code refactoring we need to do for Paracrawl 2 and 3. I would like to do it as soon as possible for users convenience.

Regarding dependency issues, I created a new 'venv' environment and yes, it looks like some dependency versions were deleted from pip or have external issues. As I haven't created any other environment in Python 3.6 since March release I didn't noticed it.

Both tensorflow 1.12.1 and urllib3 1.25 issue should be fixed with latest commit. Also, I tested with latest Anaconda and it works if you previously install datrie with conda. I will add this to the installation instructions.

About tests, given that they are most of the time broken as we are still continuously developing and producing changes to the output, we only have a "test" set that can be run with "run-tests.sh" script, downloading all it needs. It is composed of three copy-left sites crawl in WARC format ( https://github.com/bitextor/bitextor-data/releases/tag/bitextor-warc ) that are processed by Bitextor using most frequent module combination, and they are compared to a result that I manually checked, using it as ground truth. Or at least, that differs from previous version I checked but improved it somehow, sometimes replacing it as the new ground truth. This is not fast and definetely not automatic, but it is the only we have for now.

This is something we will need to work on towards CI, but still, lots of work for basic Bitextor needs to be done first.

If everything is fixed on your side, please, close this issue.

**Comment by wwaites** [comment]
Thanks, @lpla, for the quick response. I think this issue can be closed, but there are some issues still with the `run-tests.sh` both broadly as you describe, and specifically -- I cannot get them to run properly due to what seems to be a problem with the configuration file. I'll re-open #92 to continue the discussion there.
